server/src/main/kotlin/org/jetbrains/bsp/bazel/server/bep/BepServer.java

```diff
@@ -156,9 +170,35 @@ private void consumeFinishedEvent(BuildEventStreamProtos.BuildFinished buildFini
     StatusCode exitCode = ExitCodeMapper.mapExitCode(buildFinished.getExitCode().getCode());
     TaskFinishParams finishParams = new TaskFinishParams(startedEvents.pop().getKey(), exitCode);
     finishParams.setEventTime(buildFinished.getFinishTimeMillis());
+
+    if (target.isPresent()) {
+      finishParams.setDataKind(TaskFinishDataKind.COMPILE_REPORT);
+      var isSuccess = exitCode.getValue() == 1;
+      var errors = isSuccess ? 0 : 1;
+      CompileReport report = new CompileReport(target.get(), errors, 0);
+      finishParams.setData(report);
+      if (isSuccess) {
+        clearDiagnostics();
+      }
+    }
     bspClient.onBuildTaskFinish(finishParams);
   }
 
+
+  // clear former diagnostics by publishing an empty array of diagnostics
+  // why we do this on `task_finished` instead of `action_completed` or `target_complete`?
+  // - `action_completed` won't be published on build success for a target,
+  // - `target_complete` can be result of build aspect, not compilation.
+  // https://github.com/bazelbuild/bazel/blob/d43737f95d28789bb2d9ef2d7f62320e9a840ab0/src/main/java/com/google/devtools/build/lib/buildeventstream/proto/build_event_stream.proto#L157-L160
+  private void clearDiagnostics() {
+    if (target.isPresent()) {
+      var targetId = target.get().getUri();
+      var label = targetId.startsWith("@") ? targetId.substring(1) : targetId;
```

Comment:
for bzl mod project it's possible to have `@@` as well, might it be a problem?

Reply:
Turns out we can't clear diagnostics in `taskFinished`, there still were some false positives. I've added it back to `completedEvent`, but with an additional check to ensure we only clear diagnostics on successful compilation